### DISCUSSION TITLE
What ten questions would you ask a professor who might start a new class on performance engineering?

### DISCUSSION THREAD

**Discussion by behoppe** [OpenCilk/opencilk-project#94] [General]
Hi OpenCilk Community,

John Owens and Charles have drafted the following questions to guide some one-hour interviews they are scheduling over the coming weeks with a few carefully selected professors in CS and similar departments. The goal of each interview is to understand their interest in teaching performance engineering and what the specific obstacles might be. By having these discussions, we hope to promote performance engineering and organize ways to address the obstacles to teaching. We would appreciate your feedback on these questions. Thank you!

DRAFT INTERVIEW QUESTIONS:

Q1: Have you ever introduced a new course at your university? What led you to make the decision to introduce a new course?

Q2: At your university, what are the biggest obstacles to introducing a new course?

Q3: When introducing a new course at your university, is it more common to be assigned a topic/focus for the course or for you to choose the content of that course?

Q4: Assuming you have an ability to choose the content of a new course, how important are the following factors for you?
Projected level of the course (lower division undergrad, upper division undergrad, beginning graduate, advanced graduate)
1. How it fits into your university’s curriculum in terms of course material
2. Demand for the course material from students
3. Demand for the course material from industry
4. Availability of existing course materials
5. Availability of a textbook
6. Availability of infrastructure to teach the course (e.g., servers or cloud computers), and $ to purchase hardware/IT support/cloud computing time
7. This course is taught at many other universities
8. This course is NOT taught at many other universities

Q5: If we were specifically talking about a performance engineering course, would your answers to the important factors change?

Q6: If you were asked to prepare a term-length performance engineering course for advanced undergrads/beginning grad students, what topics would you expect to include?

Q7: If experienced teachers of this course were to offer in-person or remote training/overview/etc. to instructors interested in teaching this class—a weeklong workshop in person over the summer, for instance, or a 2-day Zoom session—would that be interesting to you? What training would be interesting for you? Have you ever participated in such training before? What was good/bad about it?

Q8: For each of the following topics, how comfortable are you teaching this today / would this make you nervous if it was in a course you were teaching / would you be interested in teaching this topic as a unit in a course you already teach?
1. Principles of (C) code optimization
2. Computer architecture basics
3. C compiler operation and code generation
4. Compiler optimizations
5. Parallel computing concepts and techniques (races, synchronization, threads, etc.)
6. Multicore programming
7. Algorithms and analysis

Q9: What is missing from the above list that you think would be an interesting topic in a performance engineering class?

**Comment by dcurtisatmit** [comment]
The questions seem to be divided into two groups:              
        1. How is a new course created at a particular institution (Q1-Q4)
        2. About a performance engineering course in particular    
I suggest that you alert the interviewee to these two collections of
questions.

Further, I have heard comments that professors want to teach a small unit
on performance engineering/opencilk/multicore programming within another
(already existing) course.  These questions do not seem to address that     
option?!  Of course, you can keep this in your back pocket as an alternative...

The hardware platform available for teaching a perf eng course is a
significant question (not addressed in the questions).

I recall from John Owens' experience, that the teaching assistants had    
to do a lot of work to make the course work (and John didn't write a line
of code!).  The availability of TAs and zoom meetings for them are also          
reasonable to discuss with interviewees.

**Reply by behoppe** [reply]
Thank you, @dcurtisatmit, for your comments. In short, yes, I agree :-)

John and Charles could respond more, but I know John especially is wanting to promote a coherent vision of performance engineering, not to invite feedback that might dilute that. So the focus on 6.172 is intentional. I think it's worth considering if/how we cultivate small units "in parallel" with this focus, because the small-unit approach has come up quite a bit IMO.

The hardware platform is supposed to be addressed by Q4(6). Maybe that should be more prominent?

I welcome additional comments from you and others. Thanks!

**Comment by angelee** [comment]
I don't have comments to add at the moment, but during the meeting, I was
asked to think of names of people that I know who might be interesting to
interview.  A couple names come to mind (that are likely not on Charles'
radar):

Michael Bond in Ohio State (PL / System)
Joseph Devietti at UPenn (Architecture / System)

I-Ting Angelina Lee
Associate Professor, CSE, Washington University in St. Louis
https://www.cse.wustl.edu/~angelee/home_page/
